### PR TITLE
Fix how we get a quantization method from config

### LINKED ISSUE #25550: Quantization model initialization routine tries to use quantization_config as if it is a dict
The whole problem is here https://github.com/huggingface/transformers/blob/36f183ebab1261b388739d628aaa0b4150068df0/src/transformers/modeling_utils.py#L2389C34-L2389C34

One part of quantized model initialization tries to use `.get(...)` method of `quantization_config`. Presumably, this was made with the assumption that `quantization_config` would be a dict loaded from some json. However, this is not always the case.

For some reason it is not caught by tests but is was caught by me during training with modified [run_classification.py](https://github.com/huggingface/transformers/blob/36f183ebab1261b388739d628aaa0b4150068df0/examples/pytorch/text-classification/run_classification.py).

Here is customized model_init function:
```python
def model_init():
      if not model_args.use_lora and model_args.n_bits in {4, 8}:
          raise ValueError("4 and 8bit modes can only be used with LoRA")

      if model_args.n_bits not in {4, 8, 16, 32}:
          raise ValueError("Only 4, 8, 16, and 32bit modes are supported")

      model_kwargs = {}
      if model_args.n_bits == 4:
          model_kwargs['quantization_config'] = BitsAndBytesConfig(
              load_in_4bit=True,
              bnb_4bit_compute_dtype=torch.bfloat16,
              bnb_4bit_quant_type='nf4',
              bnb_4bit_use_double_quant=True
          )
          model_kwargs['torch_dtype'] = torch.bfloat16
      elif model_args.n_bits == 8:
          model_kwargs['quantization_config'] = BitsAndBytesConfig(
              load_in_8bit=True
          )

      model = AutoModelForSequenceClassification.from_pretrained(
          model_args.model_name_or_path,
          from_tf=bool(".ckpt" in model_args.model_name_or_path),
          config=config,
          cache_dir=model_args.cache_dir,
          revision=model_args.model_revision,
          token=model_args.token,
          trust_remote_code=model_args.trust_remote_code,
          ignore_mismatched_sizes=model_args.ignore_mismatched_sizes,
          **model_kwargs
      )

      if model_args.use_lora:
          if model_args.n_bits in {4, 8}:
              model = peft.prepare_model_for_kbit_training(model)

          lora_config = peft.LoraConfig(
              task_type=peft.TaskType.SEQ_CLS,
              r=128, lora_alpha=64,
              lora_dropout=0.05,
              inference_mode=False
          )
          model = peft.get_peft_model(model, lora_config)
          model.print_trainable_parameters()

      return model
```

And here is the stack trace:
```python
Traceback (most recent call last):
  File "/workspace/notebooks/wmt23/run_regression.py", line 786, in <module>
    main()
  File "/workspace/notebooks/wmt23/run_regression.py", line 724, in main
    train_result = trainer.train(resume_from_checkpoint=checkpoint)
  File "/usr/local/lib/python3.10/dist-packages/transformers/trainer.py", line 1509, in train
    self.model = self.call_model_init(trial)
  File "/usr/local/lib/python3.10/dist-packages/transformers/trainer.py", line 1239, in call_model_init
    model = self.model_init()
  File "/workspace/notebooks/wmt23/run_regression.py", line 554, in model_init
    model = AutoModelForSequenceClassification.from_pretrained(
  File "/usr/local/lib/python3.10/dist-packages/transformers/models/auto/auto_factory.py", line 516, in from_pretrained
    return model_class.from_pretrained(
  File "/usr/local/lib/python3.10/dist-packages/transformers/modeling_utils.py", line 2389, in from_pretrained
    quantization_method_from_config = config.quantization_config.get(
AttributeError: 'BitsAndBytesConfig' object has no attribute 'get'
```

### PR DESCRIPTION
**Pull request by Rexhaif**
# What does this PR do?

Fixes #25550 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by ArthurZucker** [comment]
cc @younesbelkada and @SunMarc

File: src/transformers/modeling_utils.py [hunk 1]
```diff
@@ -2386,9 +2386,12 @@ def from_pretrained(
         quantizer = None
         quantization_method_from_config = None
         if hasattr(config, "quantization_config"):
-            quantization_method_from_config = config.quantization_config.get(
-                "quant_method", QuantizationMethod.BITS_AND_BYTES
-            )
+            if hasattr(config.quantization_config, "get"):
```
Comment (younesbelkada) [inline src/transformers/modeling_utils.py:2389]:
```suggestion
            if isinstance(config.quantization_config, dict):
```

**Review by younesbelkada — COMMENTED** [review]
The changes look good! I made a suggestion, what do you think?

File: src/transformers/modeling_utils.py (hunk 1 above)
Comment (Rexhaif) [reply src/transformers/modeling_utils.py:2389]: Yes, that is a lot more verbose

**Review by younesbelkada — APPROVED** [review]
Thanks ! cc @SunMarc

**Comment by SunMarc** [comment]
Left a [comment](https://github.com/huggingface/transformers/issues/25550#issuecomment-1682762290) on the issue. If we have a `quantization_config` field stored on the config, it means that we are quantized weights in the state_dict. The `quantization_config` has to be `dict` because we still need to infer which quantization method was used through the `quantization_method` field. The issue is probably somewhere else.

**Review by SunMarc — COMMENTED** [review]
Thanks for the PR. Left a [comment](https://github.com/huggingface/transformers/issues/25550#issuecomment-1682762290) on the issue. If we have a `quantization_config` field stored on the config, it means that we have quantized weights in the state_dict. The `quantization_config` has to be `dict` because we still need to infer which quantization method was used through the `quantization_method` field. The issue is probably somewhere else.

**Pull request closed without merging**